refactor(api, robot-server): Refactor ProtocolRunner per protocol type

TamarZanzouri opened the pull request:
# Overview

Closes RSS-200.
create an abstract interface for `ProtocolRunner` and implement per `JsonRunner` `PythonAndLegacyRunner` and `LiveRunner`.

Also fixes a bug that was causing live http runs to finish executing the run after the first command (RSS-214).

# Test Plan

- Upload python 2.14, json v6, legacy python protocols and make sure you're able to analyze protocols, run LPC and run protocols. Check that:
    - all runs are created correctly
    - can be started, paused and stopped
    - run results are correct
- (Optional)Test a LiveRunner: Start an empty run and post protocol-intent commands. Make sure all commands are run successfully. Also test pausing and stopping the protocol.

# Changelog

- Refactor ProtocolRunner to be an abstract and implement per `JsonRunner` `PythonAndLegacyRunner` and `LiveRunner`.
- `create_simulating_runner` is now constructed per file type. 
- Change `ProtocolAnalyzer` to create on the fly the simulating runner instead of injecting it. 
- Changed tests that were using an empty run.

# Review requests

- I (Sanniti) have tested the changes on the dev server but it's important to test on a real robot (ot2/ot3 either works) since run flow control actions can't be tested on the dev server.
- We have talked about how having a runner that can allow execution of batch commands could be useful for future features. I agree that it would be useful but the way the engine & run is architected currently, it would need a more detailed change to the architecture (more importantly, careful testing) than is strictly necessary for run split. So I have not attempted to do that in this PR. We will be able to make that change in the future when the feature is on the roadmap.
- Currently we are setting run_started_at (also in the run response) when there is a play action. with this change we will no longer set this value for setup commands/post commands. this is the only real noticeable change -> might want to check with UI folks. they should use createdAt instead for this case. 

# Risk assessment

Medium-high. need to make sure nothing is effected from the refactor for all protocol types and for posting commands/runs with no protocol.

TamarZanzouri committed:
WIP: created an abstract interface for ProtocolRunner. started JsonRunner

TamarZanzouri committed:
WIP create_protocol_runner

TamarZanzouri committed:
test json_runner and finished logic for json_runner

TamarZanzouri committed:
WIP python and legacy runner

TamarZanzouri committed:
python and legacy runner implementation

TamarZanzouri committed:
create_simulating_runner.py

TamarZanzouri committed:
fixed robot server dependencies.py

TamarZanzouri committed:
MaintenanceRunner

TamarZanzouri committed:
wip robot-server

TamarZanzouri committed:
changed create_simulating_runner to monkeypatch

TamarZanzouri committed:
removed set_run_started_at and fixed test failing on started_at

TamarZanzouri committed:
rollback depends for protocol_analyzer.py and moved runner creation into analyze

TamarZanzouri commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,37 +266,156 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
+        protocol_source: Optional[ProtocolSource] = None,
+    ) -> ProtocolRunResult:
+        """Run a given protocol to completion."""
+        # TODO(mc, 2022-01-11): move load to runner creation, remove from `run`
```
can totally do this for this pr

TamarZanzouri commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,37 +266,156 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
-            )
+        This value is latched; once it is True, it will never become False.
+        """
+        return self._protocol_engine.state_view.commands.has_been_played()
```
for maintenance run should probably be true all the time.

TamarZanzouri replied:
no-op instead.

TamarZanzouri committed:
reorganize prtocol_runner tests and text fixes

TamarZanzouri commented on robot-server/tests/integration/http_api/runs/test_pause_run_not_started.tavern.yaml:
```diff
@@ -37,3 +51,13 @@ stages:
           - id: 'RunActionNotAllowed'
             title: 'Run Action Not Allowed'
             detail: 'Cannot pause a run that is not running.'
+
```
investigate why this is hanging without the stop

TamarZanzouri replied:
this is hanging on edge too

TamarZanzouri committed:
removed task_queue from maintenance runner

TamarZanzouri committed:
Merge branch 'edge' into RSS-200-refactor-protocol-runner

TamarZanzouri committed:
linting

TamarZanzouri committed:
fixed failing test in actions router

TamarZanzouri committed:
fixed maintenance test with task runner

TamarZanzouri commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,37 +266,151 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
+        protocol_source: Optional[ProtocolSource] = None,
+    ) -> ProtocolRunResult:
+        """Run a given protocol to completion."""
+        if protocol_source:
```
removed this. will never be none empty

TamarZanzouri commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,37 +266,151 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
+
+        run_data = self._protocol_engine.state_view.get_summary()
+        commands = self._protocol_engine.state_view.commands.get_all()
+        return ProtocolRunResult(commands=commands, state_summary=run_data)
```
should not be a ProtocolRunResult

TamarZanzouri commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,37 +266,151 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
+
+    async def stop(self) -> None:
+        """Stop (cancel) the run."""
+        if self.was_started():
```
for maintenance runs we should only finish bc was_started is related to protocol runs only

TamarZanzouri commented on api/tests/opentrons/protocol_runner/test_protocol_runner.py:
```diff
@@ -97,29 +104,147 @@ def use_mock_extract_labware_definitions(
+    )
+
+
+async def test_create_protocol_runner(
```
paramerarize this

TamarZanzouri commented on api/tests/opentrons/protocol_runner/test_protocol_runner.py:
```diff
@@ -512,3 +601,102 @@ async def test_load_legacy_json(
+        False, True
+    )
+
+    assert legacy_subject.was_started() is False
```
this is wrong. need to fix this test to test real cases

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -34,9 +38,7 @@ class ProtocolRunResult(NamedTuple):
-# TODO(mc, 2022-01-11): this class has become bloated. Split into an abstract
-# interfaces and several concrete implementations per protocol type
-class ProtocolRunner:
+class ProtocolRunner(TypingProtocol):
```
Instead of `TypingProtocol`, should `ProtocolRunner` be an `abc.ABC`? As far as I can tell, nothing outside of this module is intended to implement this interface.

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -34,9 +38,7 @@ class ProtocolRunResult(NamedTuple):
     state_summary: StateSummary
 
 
-# TODO(mc, 2022-01-11): this class has become bloated. Split into an abstract
-# interfaces and several concrete implementations per protocol type
```
:)

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -48,22 +50,50 @@ class ProtocolRunner:
+    async def stop(self) -> None:
+        """Stop (cancel) the run."""
+
+    async def run(
+        self,
+        protocol_source: Optional[ProtocolSource] = None,
+    ) -> ProtocolRunResult:
+        """Run a given protocol to completion."""
```
It seems like this should now be a `@classmethod` or something on `PythonAndLegacyRunner` and `JsonRunner`?

TamarZanzouri replied:
per discussion with Max. we should change `load` to be a class method that returns the runner instance and remove `protocol_source` arg from this method. not for this pr. I will leave a TODO comment. Thank you @SyntaxColoring

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -48,22 +50,50 @@ class ProtocolRunner:
+
+
+class PythonAndLegacyRunner(ProtocolRunner):
+    """Protocol runner implementation for python protocols and legacy protocols."""
```
```suggestion
    """Protocol runner implementation for Python protocols, and JSON protocols ≤v5."""
```

Just to be super clear about what we mean by "legacy."

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,37 +266,151 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
-            broker=broker,
-            equipment_broker=equipment_broker,
+    async def load(self, protocol_source: ProtocolSource) -> None:
+        """Load a ProtocolSource into managed ProtocolEngine.
```
Can we not duplicate these docstrings?

All of these classes implement a common interface, and furthermore, that interface is defined in this same file. I think the interface class should have the docstrings as the single source of truth. The implementation classes should only have docstrings if they provide meaningful additional details on top of that.

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -142,7 +188,48 @@ async def run(
+        self._hardware_api = hardware_api
+        self._json_file_reader = json_file_reader or JsonFileReader()
+        self._json_translator = json_translator or JsonTranslator()
+        # TODO(mc, 2022-01-11): replace task queue with specific implementations
+        # of runner interface
```
Not for this PR, but I hope we can keep moving towards resolving this `# todo` to get rid of `TaskQueue`. 

`TaskQueue` is basically a reimplementation of standard Python `try`/`finally` control flow, in the form of methods, so we could use `decoy.verify()` on them. In my opinion, this was not a good tradeoff. Whatever architectural benefits we gained from testability, we lost because of the confusion and extra indirection that it adds.

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,37 +266,151 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
+                legacy_executor=legacy_executor,
+            )
+
+    return MaintenanceRunner(
+        protocol_engine=protocol_engine,
+        hardware_api=hardware_api,
+    )
```
As written, I think we will use `MaintenanceRunner` objects for *all* HTTP-driven runs, not just ones used for robot maintenance. In other words, if I did `POST /runs {"data": {}}` and then `POST`ed a bunch of liquid handling commands to it, I think it would go through `MaintenanceRunner`.

Two things worry me about this:

1. The name of the class becomes somewhat misleading. Maybe `LiveRunner` or `LiveAndMaintenanceRunner`?
2. I *think* it will break public usage of the HTTP API because `play`/`pause` will raise `NotImplementedError` where they were working before. I asked about this in [Slack](https://opentrons.slack.com/archives/G01LDD5U3C6/p1679691034929149).

SyntaxColoring reviewed (commented):
Looking great so far!

TamarZanzouri committed:
Update api/src/opentrons/protocol_runner/protocol_runner.py

Co-authored-by: Max Marrone <max@opentrons.com>

TamarZanzouri committed:
pr feedback

TamarZanzouri committed:
changed ProtocolRunResult to RunnerRunResult

TamarZanzouri commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -27,43 +31,75 @@
 
 
-class ProtocolRunResult(NamedTuple):
+class RunnerRunResult(NamedTuple):
```
do you have a better name?

sanni-t replied:
Why not keep it `ProtocolRunResult`?

TamarZanzouri committed:
added linting ignore

TamarZanzouri commented on api/tests/opentrons/protocol_runner/test_protocol_runner.py:
```diff
@@ -97,57 +104,144 @@ def use_mock_extract_labware_definitions(
+            json_file_reader=json_file_reader,
+            json_translator=json_translator,
+        ),
+        runner_type,  # type: ignore[arg-type]
```
not sure if this is ok. the tests are passing properly but I cant resolve the linting errors. any idea's?

SyntaxColoring replied:
Yeah, we shouldn't need this `# type: ignore`. I think we can fix this by declaring the `runner_type` argument as `typing.Type[RunnerType]`, not `RunnerType`.

TamarZanzouri committed:
fixed g-code testing

TamarZanzouri committed:
g-code testing fix

TamarZanzouri committed:
linting

TamarZanzouri committed:
linting

TamarZanzouri committed:
added pe.play for run live commands

TamarZanzouri commented:
Tested on an OT-2 with json, python, legacy protocols. tested with live protocol commands. works as expected.

TamarZanzouri committed:
added tavern test for posting protocol commands

sanni-t commented on api/src/opentrons/protocol_runner/create_simulating_runner.py:
```diff
@@ -1,20 +1,23 @@
-    """Create a ProtocolRunner wired to a simulating HardwareControlAPI.
+async def create_simulating_runner(
+    robot_type: RobotType, protocol_config: ProtocolConfig
+) -> AbstractRunner:
```
I think this should return a Union of available runners instead of Abstract Runner.

sanni-t commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -27,43 +31,76 @@
-    The ProtocolRunner is primarily responsible for feeding a ProtocolEngine
+    The AbstractRunner is primarily responsible for feeding a ProtocolEngine
     with commands and control signals. These commands and signals are
     generated by protocol files, hardware signals, or externally via
```
```suggestion
    generated by JSON/Python protocol files, hardware signals, or externally via
```

sanni-t commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -27,43 +31,76 @@
-# TODO(mc, 2022-01-11): this class has become bloated. Split into an abstract
-# interfaces and several concrete implementations per protocol type
-class ProtocolRunner:
+class AbstractRunner(ABC):
```
```suggestion
class AbstractProtocolRunner(ABC):
```
I think this will make naming related classes/ methods a lot easier.

sanni-t commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,36 +251,132 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
+            await self._protocol_engine.finish(
+                drop_tips_and_home=False,
+                set_run_status=False,
+            )
```
I think we can implement these in the abstract itself.

sanni-t commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,36 +251,132 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
-            self._protocol_engine.add_plugin(
-                LegacyContextPlugin(broker=broker, equipment_broker=equipment_broker)
+    async def load(self, protocol_source: ProtocolSource) -> None:  # noqa: D102
+        raise NotImplementedError(
```
Raise a more appropriate error

sanni-t commented on api/src/opentrons/protocol_runner/create_simulating_runner.py:
```diff
@@ -1,20 +1,23 @@
-async def create_simulating_runner(robot_type: RobotType) -> ProtocolRunner:
-    """Create a ProtocolRunner wired to a simulating HardwareControlAPI.
+async def create_simulating_runner(
+    robot_type: RobotType, protocol_config: ProtocolConfig
```
See if accepting a `protocol_type` makes more sense here.

sanni-t replied:
Adding a TODO to address later.

sanni-t committed:
addressed PR comments, fixed live runner bug, updated tests

sanni-t committed:
updated tests, fixed linter errors

b-cooper commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,36 +233,102 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
+
+
+class LiveRunner(AbstractRunner):
+    """Protocol runner implementation for setup commands."""
```
Is this docstring accurate?  It seems a bit unclear. What is a "setup" command in this context?

sanni-t replied:
Good catch

sanni-t replied:
```suggestion
    """Protocol runner implementation for live http protocols."""
```

sanni-t committed:
Merge branch 'edge' into RSS-200-refactor-protocol-runner

sanni-t committed:
correct docstring

b-cooper reviewed (approved):
Tested via App branch build, successfully analyzed python and json protocols. ✅

sanni-t commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -27,43 +31,76 @@
+            await self._protocol_engine.stop()
+        else:
+            await self._protocol_engine.finish(
+                drop_tips_and_home=False,
```
Investigate if this should be actually true.

sanni-t replied:
Our current code in edge does tip drop only after a clean protocol run finish and not in `runner.stop` (run cancellation) so this keeps that behavior as is and is correct.

sanni-t commented on api/tests/opentrons/protocol_runner/smoke_tests/test_protocol_runner.py:
```diff
@@ -38,14 +43,17 @@ async def test_runner_with_python(
     tempdeck_v1_def: ModuleDefinition,
 ) -> None:
-    """It should run a Python protocol on the ProtocolRunner."""
+    """It should run a Python protocol on the AbstractRunner."""
```
```suggestion
    """It should run a Python protocol on the PythonAndLegacyRunner."""
```

sanni-t commented on api/tests/opentrons/protocol_runner/smoke_tests/test_protocol_runner.py:
```diff
@@ -27,8 +27,13 @@
+from opentrons.protocols.api_support.types import APIVersion
 
 
 # TODO (tz, 6-17-22): API version 3.x in-development.
```
Check what's going on here. Why's this xfailing?

sanni-t commented on api/tests/opentrons/protocol_runner/smoke_tests/test_protocol_runner.py:
```diff
@@ -1,6 +1,6 @@
-"""Smoke tests for the ProtocolRunner and ProtocolEngine classes.
+"""Smoke tests for the AbstractRunner and ProtocolEngine classes.
```
Add LiveRunner test if it doesn't exist.

sanni-t replied:
There's an integration test for LiveRunner which covers the requirement.

sanni-t commented on api/tests/opentrons/protocol_runner/test_legacy_command_mapper.py:
```diff
@@ -1,4 +1,4 @@
-"""Tests for the ProtocolRunner's LegacyContextPlugin."""
+"""Tests for the AbstractRunner's LegacyContextPlugin."""
```
```suggestion
"""Tests for the PythonAndLegacyRunner's LegacyCommandMapper."""
```

sanni-t commented on api/tests/opentrons/protocol_runner/test_legacy_context_plugin.py:
```diff
@@ -1,4 +1,4 @@
-"""Tests for the ProtocolRunner's LegacyContextPlugin."""
+"""Tests for the AbstractRunner's LegacyContextPlugin."""
```
```suggestion
"""Tests for the PythonAndLegacyRunner's LegacyContextPlugin."""
```

sanni-t commented on api/tests/opentrons/protocol_runner/test_protocol_runner.py:
```diff
@@ -1,8 +1,9 @@
-"""Tests for the ProtocolRunner class."""
+"""Tests for the AbstractRunner class."""
```
```suggestion
"""Tests for the PythonAndLegacyRunner, JsonRunner & LiveRunner classes."""
```

sanni-t commented on api/tests/opentrons/protocol_runner/test_protocol_runner.py:
```diff
@@ -97,57 +105,145 @@ def use_mock_extract_labware_definitions(
+
+
+@pytest.fixture
+def legacy_subject(
```
Rename to `python_runner_subject`?

sanni-t commented on api/tests/opentrons/protocol_runner/test_protocol_runner.py:
```diff
@@ -97,57 +105,145 @@ def use_mock_extract_labware_definitions(
-    """Get a ProtocolRunner test subject with mocked dependencies."""
-    return ProtocolRunner(
+) -> PythonAndLegacyRunner:
+    """Get a JsonRunner test subject with mocked dependencies."""
```
```suggestion
    """Get a PythonAndLegacyRunner test subject with mocked dependencies."""
```

sanni-t commented on robot-server/robot_server/protocols/protocol_analyzer.py:
```diff
@@ -1,7 +1,7 @@
 import logging
 
-from opentrons.protocol_runner import ProtocolRunner
+import opentrons.protocol_runner as protocol_runner
```
```suggestion
from opentrons import protocol_runner
```

sanni-t commented on robot-server/robot_server/runs/engine_store.py:
```diff
@@ -27,10 +33,10 @@ class EngineConflictError(RuntimeError):
 
 class RunnerEnginePair(NamedTuple):
-    """A stored ProtocolRunner/ProtocolEngine pair."""
+    """A stored AbstractRunner/ProtocolEngine pair."""
```
```suggestion
    """A stored Runner/ProtocolEngine pair."""
```

sanni-t commented on robot-server/robot_server/runs/engine_store.py:
```diff
@@ -132,16 +138,25 @@ async def create(
+                protocol is not None
+            ), "A Python or JSON protocol should have a protocol source file."
             await runner.load(protocol.source)
+        else:  # Why does linter not like `else:`..?
```
I think this is outdated comment.

sanni-t reviewed (commented):
Comments from live review.

SyntaxColoring commented on api/tests/opentrons/protocol_runner/smoke_tests/test_legacy_command_mapper.py:
```diff
@@ -32,7 +33,10 @@ async def simulate_and_get_commands(protocol_file: Path) -> List[commands.Comman
+    subject = await create_simulating_runner(
+        robot_type="OT-2 Standard",
+        protocol_config=PythonProtocolConfig(api_version=APIVersion(2, 13)),
+    )
```
```suggestion
    subject = await create_simulating_runner(
        robot_type="OT-2 Standard",
        protocol_config=protocol_source.config,
    )
```

SyntaxColoring commented on api/tests/opentrons/protocol_runner/smoke_tests/test_legacy_custom_labware.py:
```diff
@@ -52,7 +53,10 @@ async def test_legacy_custom_labware(custom_labware_protocol_files: List[Path])
+    subject = await create_simulating_runner(
+        robot_type="OT-2 Standard",
+        protocol_config=PythonProtocolConfig(api_version=APIVersion(2, 13)),
+    )
```
```suggestion
    subject = await create_simulating_runner(
        robot_type="OT-2 Standard",
        protocol_config=protocol_source.config,
    )
```

SyntaxColoring commented on api/tests/opentrons/protocol_runner/smoke_tests/test_legacy_module_commands.py:
```diff
@@ -60,7 +64,10 @@ async def test_runner_with_modules_in_legacy_python(
+    subject = await create_simulating_runner(
+        robot_type="OT-2 Standard",
+        protocol_config=PythonProtocolConfig(api_version=APIVersion(2, 13)),
+    )
```
```suggestion
    subject = await create_simulating_runner(
        robot_type="OT-2 Standard",
        protocol_config=protocol_source.config,
    )
```

SyntaxColoring commented on api/tests/opentrons/protocol_runner/smoke_tests/test_protocol_runner.py:
```diff
@@ -38,14 +43,17 @@ async def test_runner_with_python(
+    subject = await create_simulating_runner(
+        robot_type="OT-2 Standard",
+        protocol_config=PythonProtocolConfig(api_version=APIVersion(2, 14)),
+    )
```
```suggestion
    subject = await create_simulating_runner(
        robot_type="OT-2 Standard",
        protocol_config=protocol_source.config,
    )
```

SyntaxColoring commented on api/tests/opentrons/protocol_runner/smoke_tests/test_protocol_runner.py:
```diff
@@ -104,14 +112,16 @@ async def test_runner_with_python(
-    subject = await create_simulating_runner(robot_type="OT-2 Standard")
+    subject = await create_simulating_runner(
+        robot_type="OT-2 Standard", protocol_config=JsonProtocolConfig(schema_version=6)
+    )
```
```suggestion
    subject = await create_simulating_runner(
        robot_type="OT-2 Standard", protocol_config=protocol_source.config
    )
```

SyntaxColoring commented on api/tests/opentrons/protocol_runner/smoke_tests/test_protocol_runner.py:
```diff
@@ -163,14 +173,17 @@ async def test_runner_with_json(json_protocol_file: Path) -> None:
-    subject = await create_simulating_runner(robot_type="OT-2 Standard")
+    subject = await create_simulating_runner(
+        robot_type="OT-2 Standard",
+        protocol_config=PythonProtocolConfig(api_version=APIVersion(2, 13)),
```
```suggestion
        protocol_config=protocol_source.config,
```

SyntaxColoring commented on api/tests/opentrons/protocol_runner/smoke_tests/test_protocol_runner.py:
```diff
@@ -220,14 +233,16 @@ async def test_runner_with_legacy_python(legacy_python_protocol_file: Path) -> N
 
-    subject = await create_simulating_runner(robot_type="OT-2 Standard")
+    subject = await create_simulating_runner(
+        robot_type="OT-2 Standard", protocol_config=JsonProtocolConfig(schema_version=5)
```
```suggestion
        robot_type="OT-2 Standard", protocol_config=protocol_source.config
```

SyntaxColoring commented on robot-server/robot_server/protocols/router.py:
```diff
@@ -46,8 +46,8 @@
     get_protocol_store,
     get_analysis_store,
-    get_protocol_analyzer,
     get_protocol_directory,
+    get_protocol_analyzer,
```
Accidental diff?

sanni-t replied:
🤷🏻‍♀️

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,36 +233,102 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
-            context=context,
-        )
+
+RunnerType = Union[PythonAndLegacyRunner, JsonRunner, LiveRunner]
```
No action needed for this PR because I think you said once that this was an existing Opentrons convention (?), but I think `RunnerType` is a confusing thing to name a union. I would do something like `AnyRunner`.

I feel like `FooType` is more conventionally used for things like:

`FooType = typing.TypeVar[...]`

Or:

`FooType = typing.Type[...]`

SyntaxColoring commented on robot-server/robot_server/runs/engine_store.py:
```diff
@@ -132,16 +138,25 @@ async def create(
+            ), "A Python or JSON protocol should have a protocol source file."
             await runner.load(protocol.source)
+        else:  # Why does linter not like `else:`..?
+            runner.set_task_queue_wait()
```
I think the name `set_task_queue_wait()` is leaking implementation details that aren't meaningful to the caller. Can we call this `load()` (for symmetry with `load(protocol.source)`) or `prepare()`?

Ideally, I don't think this separate load/prepare step would be required for *any* runner. It would be folded into the creation of the runner. But I think that's out of scope for this PR.

SyntaxColoring commented on robot-server/tests/integration/http_api/runs/test_pause_run_not_started.tavern.yaml:
```diff
@@ -5,11 +5,24 @@ marks:
 
 stages:
-  - name: Create Empty Run
+  - name: Upload a protocol
```
With these changes (making it a protocol run instead of a basic run, and adding an extra stop action at the end), I'm kind of confused about what this file is testing. The `test_name` says "make sure a non active run cannot be paused", but now it seems like we're also trying to cover something else? Whatever we're covering here, does it belong in a separate test?

sanni-t replied:
Reverted this test back to what it was. See my other comment about pre-LiveRunner changes.

SyntaxColoring commented on robot-server/tests/integration/http_api/runs/test_persistence.py:
```diff
@@ -101,9 +102,14 @@ async def test_runs_persist_via_actions_router(
     """Test that runs commands and state
     are persisted when calling play action through dev server restart."""
     client, server = client_and_server
+    await client.post_protocol([Path("./tests/integration/protocols/simple.py")])
```
A couple things about the two new `post_protocol()`s in this file:

1. Why are we adding these?
2. We're querying `get_protocols()` to get the new protocol ID. But we can get that from the `post_protocol()` response. (If we need these, in the first place.)

sanni-t replied:
Looks like these changes with including protocols in tests were made when we had split the runner into Python & JSON runners only with no LiveRunner so that there was no concept of using a run without a protocol. Now that we have a LiveRunner, I've reverted these tests back to what they were. Thanks for thoroughly checking these tests!

SyntaxColoring commented on robot-server/tests/integration/http_api/runs/test_run_queued_protocol_commands.tavern.yaml:
```diff
@@ -0,0 +1,195 @@
+      - run_server
+
+stages:
+  - name: Create Empty Run
```
```suggestion
  - name: Create empty run
```

SyntaxColoring commented on robot-server/tests/integration/http_api/runs/test_run_queued_protocol_commands.tavern.yaml:
```diff
@@ -0,0 +1,195 @@
+    response:
+      status_code: 200
+      json:
+        # Check that `data` is identical to the original POSTed run.
```
This looks like a comment copy-pasted from another Tavern test that doesn't make sense in this context.
```suggestion
```

SyntaxColoring commented on robot-server/tests/integration/http_api/runs/test_run_queued_protocol_commands.tavern.yaml:
```diff
@@ -0,0 +1,195 @@
+            completedAt: !re_search "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d+\\+\\d{2}:\\d{2}$"
+      save:
+        json:
+          load_labware_command_id: data[0].id
```
This looks unused.

```suggestion
```

SyntaxColoring commented on robot-server/tests/integration/http_api/runs/test_run_queued_protocol_commands.tavern.yaml:
```diff
@@ -0,0 +1,195 @@
+
+marks:
+  - usefixtures:
+      - run_server
```
Since #12139, we need `clean_server_state` too, to make sure this test's changes to the server don't leak into other tests.

```suggestion
      - run_server
      - clean_server_state
```

SyntaxColoring commented on robot-server/tests/integration/http_api/runs/test_run_queued_protocol_commands.tavern.yaml:
```diff
@@ -0,0 +1,195 @@
+test_name: Make sure empty run commands execute protocol commands queue
```
```suggestion
test_name: Make sure multiple commands can be queued and played on a basic run
```

SyntaxColoring commented on robot-server/tests/runs/test_engine_store.py:
```diff
@@ -140,9 +126,19 @@ async def test_clear_engine(subject: EngineStore) -> None:
+    )
+    await subject.create(
+        run_id="run-id", labware_offsets=[], protocol=protocol_resource
+    )
```
Why does this test in particular, unlike the other ones, need to provide an argument to `protocol`?

sanni-t replied:
Reverted this test back to what it was. See my other comment about pre-LiveRunner changes.

SyntaxColoring commented on robot-server/tests/runs/test_engine_store.py:
```diff
@@ -91,29 +100,6 @@ async def test_create_engine_with_labware_offsets(subject: EngineStore) -> None:
     ]
 
 
-@pytest.mark.xfail(strict=True, raises=NotImplementedError)
-async def test_create_engine_with_protocol(
-    subject: EngineStore,
-    protocol_source: ProtocolSource,
-) -> None:
-    """It should create an engine for a run with labware offsets."""
-    # TODO(mc, 2022-05-18): https://github.com/Opentrons/opentrons/pull/10170
-    raise NotImplementedError("Implement this test when JSONv6 runs are supported")
-
-    protocol = ProtocolResource(
-        protocol_id="my cool protocol",
-        protocol_key=None,
-        created_at=datetime(year=2021, month=1, day=1),
-        source=protocol_source,
-    )
-
-    await subject.create(
-        run_id="run-id",
-        labware_offsets=[],
-        protocol=protocol,
-    )
-
-
```
Hm, it looks like this whole time, this file has been missing a test that if you pass in a `protocol` argument to `subject.create()`, the subject calls `protocol_runner.load()` with the appropriate value?

I don't think it's a huge deal, but the subject *does* have some logic now around that `protocol` argument, so should we add that test here?

sanni-t replied:
I added a basic test for checking that `create` can take a protocol and produce the correct result but verifying that `protocol_runner.load` is called requires quite a bit of monkeypatching and mocking which doesn't seem worth it when that functionality is covered by integration tests that run all 3 kinds of protocols. Still, let me know if you think we should add it.

SyntaxColoring replied:
💯 your thinking makes sense to me. Thanks for checking!

SyntaxColoring commented on api/src/opentrons/protocol_runner/protocol_runner.py:
```diff
@@ -179,36 +233,102 @@ async def _load_json(self, protocol_source: ProtocolSource) -> None:
-            )
+    async def run(  # noqa: D102
+        self,
+        protocol_source: Optional[ProtocolSource] = None,
```
It's always an error, by definition, to provide a `protocol_source` to `LiveRunner.run()`, right? Ideally, this parameter wouldn't exist for `LiveRunner`, but can we at least `assert protocol_source is None` for now?

SyntaxColoring reviewed (commented):
This looks like a big improvement, thanks @TamarZanzouri and @sanni-t. It's nice that JSON runs don't have to care about dependencies for Python runs anymore, and vice-versa. Looks good to me to merge once comments are resolved.

Here are a bunch of little things, sorry for the spam!

TamarZanzouri commented on api/src/opentrons/protocol_runner/legacy_wrappers.py:
```diff
@@ -96,6 +96,8 @@ def read(
         )
 
 
+# TODO (spp, 2023-04-05): Remove 'legacy' wording since this is the context we are using
```
:thumbsup:

TamarZanzouri commented:
Totally agree with the changes added. Makes sense to move common implementations to the abstract. Thank you for fixing this @sanni-t

sanni-t committed:
Apply suggestions from code review

- Use `protocol_source.config` where possible
- Some docstring corrections
- other small misc cleanup in tests

Co-authored-by: Max Marrone <max@opentrons.com>

sanni-t committed:
fix runner smoke tests and linter errors, some renaming from PR review

sanni-t committed:
update runner typing & type naming, renamed set_task_queue_wait() to prepare()

sanni-t committed:
address all remaining review comments

SyntaxColoring reviewed (approved):
Woohoo! －＝≡ヘ(*・ω・)ノ

jbleon95 reviewed (approved):
Looks good!
